Added new cmdlets for BillingAccounts, BillingProfiles, InvoiceSections and updated Invoicing cmdlet

asarkar84 opened the pull request:
<!-- DO NOT DELETE THIS TEMPLATE -->

## Description

Added new cmdlets for BillingAccounts, BillingProfiles, InvoiceSections and updated Invoicing cmdlet
Deprecated few cmdlets which are no longer supported
Created following issue for the review: https://github.com/Azure/azure-powershell-cmdlet-review-pr/issues/647

## Checklist

- [x] I have read the [_Submitting Changes_](../blob/master/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](../blob/master/CONTRIBUTING.md)
- [x] The title of the PR is clear and informative
- [x] The appropriate `ChangeLog.md` file(s) has been updated:
    - For any service, the `ChangeLog.md` file can be found at `src/{{SERVICE}}/{{SERVICE}}/ChangeLog.md`
    - A snippet outlining the change(s) made in the PR should be written under the `## Upcoming Release` header -- no new version header should be added
- [ ] The PR does not introduce [breaking changes](../blob/master/documentation/breaking-changes/breaking-changes-definition.md)
- [x] If applicable, the changes made in the PR have proper test coverage
- [ ] For public API changes to cmdlets:
    - [ ] a cmdlet design review was approved for the changes in [this repository](https://github.com/Azure/azure-powershell-cmdlet-review-pr) (_Microsoft internal only_)
    - [ ] the markdown help files have been regenerated using the commands listed [here](../blob/master/documentation/development-docs/help-generation.md#updating-all-markdown-files-in-a-module)

adxsdkps commented:
Can one of the admins verify this patch?

VeryEarly commented on src/Billing/Billing/Az.Billing.psd1:
```diff
@@ -3,7 +3,7 @@
 # Generated by: Microsoft Corporation
 #
-# Generated on: 5/28/2020
+# Generated on: 12/11/2019
```
please do not modify this field

asarkar84 replied:
Sure!

VeryEarly commented on src/Billing/Billing/Az.Billing.psd1:
```diff
@@ -12,7 +12,7 @@
 
 # Version number of this module.
-ModuleVersion = '1.0.3'
+ModuleVersion = '2.0.0'
```
please revert back to previous version, this will be upgraded during release by Azure-PowerShell team

asarkar84 replied:
Sure!

VeryEarly commented on src/Billing/Billing/Az.Billing.psd1:
```diff
@@ -53,7 +53,7 @@ DotNetFrameworkVersion = '4.7.2'
 
 # Modules that must be imported into the global environment prior to importing this module
-RequiredModules = @(@{ModuleName = 'Az.Accounts'; ModuleVersion = '1.9.0'; })
+RequiredModules = @(@{ModuleName = 'Az.Accounts'; ModuleVersion = '1.6.5'; })
```
please revert this back to 1.9.0

asarkar84 replied:
Sure!

VeryEarly commented on src/Billing/Billing/Az.Billing.psd1:
```diff
@@ -121,7 +122,7 @@ PrivateData = @{
 
         # ReleaseNotes of this module
-        ReleaseNotes = '* Updated assembly version of consumption cmdlets'
+        ReleaseNotes = '* Update references in .psd1 to use relative path'
```
please revert this back

asarkar84 replied:
Sure!

VeryEarly commented on src/Billing/Billing/Cmdlets/BillingAccounts/GetAzureRmBillingAccount.cs:
```diff
@@ -0,0 +1,107 @@
+        public SwitchParameter IncludeAddress { get; set; }
+
+        [Parameter(Mandatory = false, HelpMessage = "Expand the billing profiles under the billing accounts.")]
+        public SwitchParameter ExpandBillingProfiles { get; set; }
```
use "ExpandBillingProfile" instead of "ExpandBillingProfiles"

asarkar84 replied:
Sure! Done.

VeryEarly commented on src/Billing/Billing/Cmdlets/BillingAccounts/GetAzureRmBillingAccount.cs:
```diff
@@ -0,0 +1,107 @@
+        public SwitchParameter ExpandBillingProfiles { get; set; }
+
+        [Parameter(Mandatory = false, HelpMessage = "Expand the billing profiles and invoice sections under the billing profiles.")]
+        public SwitchParameter ExpandInvoiceSections { get; set; }
```
use "ExpandInvoiceSections" instead of "ExpandInvoiceSections"

asarkar84 replied:
Sure! Done. Using "ExpandInvoiceSection".

VeryEarly commented on src/Billing/Billing/Cmdlets/BillingProfiles/GetAzureRmBillingProfile.cs:
```diff
@@ -0,0 +1,84 @@
+        public string BillingAccountName { get; set; }
+        
+        [Parameter(Mandatory = false, HelpMessage = "Expand the invoice sections under the billing profiles.")]
+        public SwitchParameter ExpandInvoiceSections { get; set; }
```
use "ExpandInvoiceSection" instead of "ExpandInvoiceSections"

asarkar84 replied:
Sure! Done.

VeryEarly commented on src/Billing/Billing/Properties/AssemblyInfo.cs:
```diff
@@ -46,5 +46,5 @@
 // [assembly: AssemblyVersion("1.0.2")]
-[assembly: AssemblyVersion("1.0.3")]
-[assembly: AssemblyFileVersion("1.0.3")]
+[assembly: AssemblyVersion("1.0.2")]
```
please revert to previous version

asarkar84 replied:
Sure!

asarkar84 commented:
> breaking changes are not allowed in non-breaking change release, please add these deprecated logic back, keep deprecation attribute: https://github.com/Azure/azure-powershell/blob/master/documentation/breaking-changes/breaking-changes-attribute-help.md#when-there-is-no-replacement-cmdlet

In the current version of SDK, we are not supporting the Get-AzBillingPeriod and Get-AzEnrollmentAccount cmdlets. USers were consuming these to get the UsageDetails which is part of the Consumption cmdlets. Not Consumption cmdlets has alternate ways to get the usage without the Billing Period etc.
Please make this as part of the breaking change release.

VeryEarly commented:
> > breaking changes are not allowed in non-breaking change release, please add these deprecated logic back, keep deprecation attribute: https://github.com/Azure/azure-powershell/blob/master/documentation/breaking-changes/breaking-changes-attribute-help.md#when-there-is-no-replacement-cmdlet
> 
> In the current version of SDK, we are not supporting the Get-AzBillingPeriod and Get-AzEnrollmentAccount cmdlets. USers were consuming these to get the UsageDetails which is part of the Consumption cmdlets. Not Consumption cmdlets has alternate ways to get the usage without the Billing Period etc.
> Please make this as part of the breaking change release.

Hi @asarkar84 ,

Next breaking change release should be this September, do you mean you want this PR to go with it?

asarkar84 commented:
@VeryEarly I have added back the Get-AzBillingPeriod and Get-AzEnrollmentAccount cmdlets. Removed CmdletDeprecation attribute from Get-AzEnrollmentAccount since we will keep supporting it. Keeping CmdletDeprecation attribute in Get-AzBillingPeriod cmdlet since this should be depricated in the next versions.
Please take a look.

asarkar84 commented:
> > > breaking changes are not allowed in non-breaking change release, please add these deprecated logic back, keep deprecation attribute: https://github.com/Azure/azure-powershell/blob/master/documentation/breaking-changes/breaking-changes-attribute-help.md#when-there-is-no-replacement-cmdlet
> > 
> > 
> > In the current version of SDK, we are not supporting the Get-AzBillingPeriod and Get-AzEnrollmentAccount cmdlets. Users were consuming these to get the UsageDetails which is part of the Consumption cmdlets. Now Consumption cmdlets has alternate ways to get the usage without the Billing Period etc.
> > Please make this as part of the breaking change release.
> 
> Hi @asarkar84 ,
> 
> Next breaking change release should be this September, do you mean you want this PR to go with it?

We are already in the end of August end. We are OK for releasing it with breaking change version in September. What is the exact breaking change release date?

VeryEarly commented on src/Billing/Billing/ChangeLog.md:
```diff
@@ -20,7 +20,10 @@
 
 ## Version 1.0.3
-* Updated assembly version of consumption cmdlets
+* Added `Get-AzBillingAccount` cmdlet
```
please put release notes under
## Upcoming Release

and add "* Updated assembly version of consumption cmdlets" back under ## Version 1.0.3

asarkar84 replied:
Sure! I have updated the ChangeLog and sent a new iteration.

VeryEarly commented:
> > > > breaking changes are not allowed in non-breaking change release, please add these deprecated logic back, keep deprecation attribute: https://github.com/Azure/azure-powershell/blob/master/documentation/breaking-changes/breaking-changes-attribute-help.md#when-there-is-no-replacement-cmdlet
> > > 
> > > 
> > > In the current version of SDK, we are not supporting the Get-AzBillingPeriod and Get-AzEnrollmentAccount cmdlets. Users were consuming these to get the UsageDetails which is part of the Consumption cmdlets. Now Consumption cmdlets has alternate ways to get the usage without the Billing Period etc.
> > > Please make this as part of the breaking change release.
> > 
> > 
> > Hi @asarkar84 ,
> > Next breaking change release should be this September, do you mean you want this PR to go with it?
> 
> Please note that we have added the deprecated logic back, do we still need this to be part of the breaking change release? What is the exact breaking change release date?

Hi @asarkar84 ,

This PR contains new cmdlets only now, so no need to wait for the breaking change release.

Please resolve comment and static analysis issue and ti should be good to go.

asarkar84 commented:
> > > > > breaking changes are not allowed in non-breaking change release, please add these deprecated logic back, keep deprecation attribute: https://github.com/Azure/azure-powershell/blob/master/documentation/breaking-changes/breaking-changes-attribute-help.md#when-there-is-no-replacement-cmdlet
> > > > 
> > > > 
> > > > In the current version of SDK, we are not supporting the Get-AzBillingPeriod and Get-AzEnrollmentAccount cmdlets. Users were consuming these to get the UsageDetails which is part of the Consumption cmdlets. Now Consumption cmdlets has alternate ways to get the usage without the Billing Period etc.
> > > > Please make this as part of the breaking change release.
> > > 
> > > 
> > > Hi @asarkar84 ,
> > > Next breaking change release should be this September, do you mean you want this PR to go with it?
> > 
> > 
> > Please note that we have added the deprecated logic back, do we still need this to be part of the breaking change release? What is the exact breaking change release date?
> 
> Hi @asarkar84 ,
> 
> This PR contains new cmdlets only now, so no need to wait for the breaking change release.
> 
> Please resolve comment and static analysis issue and ti should be good to go.

@VeryEarly I have fixed the security tests.

asarkar84 commented:
Hi @VeryEarly We have removed the properties DownloadUrlExpiry, Type and BillingPeriodNames from the response of Get-AzBillingInvoice cmdlet which may be considered as a breaking change. Please take a look.
We may wait till the next Breaking change release in September.

asarkar84 commented:
@VeryEarly Since we are very close to the October breaking change release, I would like to ensure that the below PR is good to go. Please take a look again and let me know if you expect me to take any action.

VeryEarly commented:
/azp run

VeryEarly commented:
/azp run

asarkar84 commented:
@VeryEarly Please merge this PR if it looks good now.

asarkar84 commented:
/azp run
